Source/Laser/LaserParticleContainer.cpp

```diff
@@ -501,14 +501,8 @@ LaserParticleContainer::Evolve (int lev,
             }
 
             if (cost) {
-                const Box& tbx = pti.tilebox();
-                wt = (amrex::second() - wt) / tbx.d_numPts();
-                Array4<Real> const& costarr = cost->array(pti);
-                amrex::ParallelFor(tbx,
-                [=] AMREX_GPU_DEVICE (int i, int j, int k) noexcept
-                {
-                    costarr(i,j,k) += wt;
-                });
+                wt = amrex::second() - wt;
+                amrex::HostDevice::Atomic::Add( &(*cost)[pti.index()], wt);
```

Comment:
Could you explain why you need an `amrex::HostDevice::Atomic::Add`?
- If this is a CPU run, do we need sth like `#omp pragma atomic`? Is this `HostDevice` really doing that?
- If this is a GPU run (and we're using timers), shouldn't we need a `cudaDeviceSynchronize` here to make sure the timer contains meaningful data?

Also, shouldn't we do this add only if we're using timer costs, and not if using heuristic?

Reply:
> Could you explain why you need an `amrex::HostDevice::Atomic::Add`?

This is prevents a race condition if we using tiling/openMP threading for the CPU case.

> * If this is a CPU run, do we need sth like `#omp pragma atomic`? Is this `HostDevice` really doing that?

A directive like that is indeed in`HostDevice`, nothing to take care of on WarpX side; see below.
```
namespace HostDevice { namespace Atomic {
    template <class T>
    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
    void Add (T* sum, T value) noexcept
    {
#if AMREX_DEVICE_COMPILE
        Gpu::Atomic::Add(sum,value);
#else
#ifdef _OPENMP
#pragma omp atomic update
#endif
        *sum += value;
#endif
    }
}}
```

> * If this is a GPU run (and we're using timers), shouldn't we need a `cudaDeviceSynchronize` here to make sure the timer contains meaningful data?

Do we need to add this?  I thought this is already present in the profiling macros(?).

> Also, shouldn't we do this add only if we're using timer costs, and not if using heuristic?

Good catch!  This extra condition is added for updating costs with timer times.

Reply:
Ok sounds good, I was confused about the name `amrex::HostDevice::Atomic::Add` but actually it makes sense and I agree it does the right thing, thanks for the code snippet.

Concerning the synchronize, I should clarify this as it is an important point: we should **not** rely on the profiler for the synchronization, because these sync may be removed at any time. **The code should be correct even without the WARPX_PROFILE cudaDeviceSynchronize()**.

Reply:
Agree, I also found the name `amrex::HostDevice::Atomic::Add` confusing.  If something like `amrex::Host::Atomic::Add` existed, that would be enough, as we're never calling the operation from the GPU, but `amrex::HostDevice::Atomic::Add` is what we have.

As for the other point, I've added device synchronizations.

Reply:
@MaxThevenet update on performance impact of the added device synchs: 

![img_scalings](https://user-images.githubusercontent.com/38045958/77571492-b3d75f80-6e8a-11ea-8009-5cd58b8e24f2.png)

not much impact, as you expected.